src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -585,17 +642,72 @@ public static string Join(string separator, params object[] values)
         }
 
         [ComVisible(false)]
-        public static String Join<T>(String separator, IEnumerable<T> values)
+        public unsafe static string Join<T>(string separator, IEnumerable<T> values)
         {
+            separator = separator ?? string.Empty;
+            fixed (char* pSeparator = &separator.m_firstChar)
+            {
+                // Defer argument validataion to the internal function
+                return JoinCore(pSeparator, separator.Length, values);
+            }
+        }
+
+        [ComVisible(false)]
+        public static String Join(String separator, IEnumerable<String> values) {
```

Comment:
Since you changed `String` -> `string` above, maybe go ahead and do the same here.